From 2a5f7b9c4d1386fb5621f74e7b43384e70ee067f Mon Sep 17 00:00:00 2001
From: Praveenkumar I <quic_ipkumar@quicinc.com>
Date: Fri, 26 Jul 2024 16:14:06 +0530
Subject: [PATCH] cpufreq: qcom-nvmem: Enable cpufreq for ipq5424

IPQ5424 have different OPPs available for the CPU based on SoC
variant. This can be determined through use of eFuse / SoftSKU
status register present in the silicon.

Added support for ipq5424 on nvmem driver which helps to determine
OPPs at runtime based on the eFuse / SoftSKU register which has
the CPU frequency PLL Lvalue.

IPQ5424 supports two limits, 1.No limit(TURBO) 2.Lvalue(NOMINAL).
Support has been added such a way that if Lvalue is present, CPU will
be limited to Nominal else Turbo will be supported.

Change-Id: Ie8acdd79f6dbb9f80ce24fffe7d32d33743add39
Signed-off-by: Praveenkumar I <quic_ipkumar@quicinc.com>
---
 drivers/cpufreq/qcom-cpufreq-nvmem.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/cpufreq/qcom-cpufreq-nvmem.c b/drivers/cpufreq/qcom-cpufreq-nvmem.c
index f6aeba35245d..e1a3774d5ed0 100644
--- a/drivers/cpufreq/qcom-cpufreq-nvmem.c
+++ b/drivers/cpufreq/qcom-cpufreq-nvmem.c
@@ -161,6 +161,10 @@ static int qcom_cpufreq_kryo_name_version(struct device *cpu_dev,
 	case QCOM_ID_IPQ9574:
 		drv->versions = 1 << (unsigned int)(*speedbin);
 		break;
+	case QCOM_ID_IPQ5424:
+	case QCOM_ID_IPQ5404:
+		drv->versions = 1 << (unsigned int)!!(*speedbin);
+		break;
 	case QCOM_ID_MSM8996SG:
 	case QCOM_ID_APQ8096SG:
 		drv->versions = 1 << ((unsigned int)(*speedbin) + 4);
@@ -371,6 +375,7 @@ static const struct of_device_id qcom_cpufreq_match_list[] __initconst = {
 	{ .compatible = "qcom,msm8996", .data = &match_data_kryo },
 	{ .compatible = "qcom,qcs404", .data = &match_data_qcs404 },
 	{ .compatible = "qcom,ipq5332", .data = &match_data_kryo },
+	{ .compatible = "qcom,ipq5424", .data = &match_data_kryo },
 	{ .compatible = "qcom,ipq8064", .data = &match_data_krait },
 	{ .compatible = "qcom,ipq9574", .data = &match_data_kryo },
 	{ .compatible = "qcom,apq8064", .data = &match_data_krait },
-- 
2.34.1

